From 1531df29a503c8300571dcc785c2ac032cf99bcc Mon Sep 17 00:00:00 2001
From: Valentine Barshak <valentine.barshak@cogentembedded.com>
Date: Tue, 19 Jun 2018 20:52:07 +0300
Subject: [PATCH] plat: renesas: rcar: Set maximum supported RPC frequency

This sets maximum supported RPC interface frequency in BL2
which should speed up image loading. We also drop the frequency
to 80MHz before switching to non-secure context in BL31 in order
to avoid possible RPC HyperFlash write issues when non-secure
RPC access is enabled (RCAR_DISABLE_NONSECURE_RPC_ACCESS = 0).

Signed-off-by: Valentine Barshak <valentine.barshak@cogentembedded.com>
---
 plat/renesas/rcar/bl31_rcar_setup.c        | 3 +++
 plat/renesas/rcar/drivers/rpc/rpc_driver.c | 6 ++++++
 2 files changed, 9 insertions(+)

diff --git a/plat/renesas/rcar/bl31_rcar_setup.c b/plat/renesas/rcar/bl31_rcar_setup.c
index a5e494e..90ebd37 100644
--- a/plat/renesas/rcar/bl31_rcar_setup.c
+++ b/plat/renesas/rcar/bl31_rcar_setup.c
@@ -240,6 +240,9 @@ void bl31_plat_runtime_setup(void)
 	/* Enable non-secure access to the RPC HyperFlash region. */
 	mmio_write_32(0xee2000b8, 0x155);
 	mmio_write_32(0xee200000, mmio_read_32(0xee200000) & 0x7fffffff);
+
+	/* Reduce RPC frequency to avoid HyperFlash write issues */
+	rpc_clk_freq_set(80);
 #endif
 	/*
 	 * Finish the use of console driver in BL31 so that any runtime logs
diff --git a/plat/renesas/rcar/drivers/rpc/rpc_driver.c b/plat/renesas/rcar/drivers/rpc/rpc_driver.c
index 1af648f..10a1f9d 100644
--- a/plat/renesas/rcar/drivers/rpc/rpc_driver.c
+++ b/plat/renesas/rcar/drivers/rpc/rpc_driver.c
@@ -28,6 +28,12 @@ static void enableRPC(void)
 
 static void setupRPC(void)
 {
+	uint32_t freq;
+
+	/* Set maximum supported frequency */
+	freq = rpc_clk_freq_max();
+	rpc_clk_freq_set(freq);
+
 	/* manual mode */
 	if ((mmio_read_32(RPC_CMNCR) & RPC_CMNCR_MD_BIT) != 0U) {
 		/* external address space read mode */
-- 
2.7.4

